Non-controversial new eslint 0.24.0 rules and cleanup up formatting

andrewdeandrade opened the pull request:
This PR adds most of the new rules in eslint 0.24.0 that are non-controversial or were already enforced in lint-trap but missing in uber/standard. A separate PR will be submitted for those rules that are deemed more controversial.

Rules that are now turned on:
- [`no-catch-shadow`](https://github.com/eslint/eslint/blob/master/docs/rules/no-catch-shadow.md)
- [`no-continue`](https://github.com/eslint/eslint/blob/master/docs/rules/no-continue.md)
- [`no-div-regex`](https://github.com/eslint/eslint/blob/master/docs/rules/no-div-regex.md)
- [`no-else-return`](https://github.com/eslint/eslint/blob/master/docs/rules/no-else-return.md)
- [`no-eq-null`](https://github.com/eslint/eslint/blob/master/docs/rules/no-eq-null.md) (probably a noop since we have strict equality required)
- [`no-extra-parens`](https://github.com/eslint/eslint/blob/master/docs/rules/no-extra-parens.md)
- [`no-inline-comments`](https://github.com/eslint/eslint/blob/master/docs/rules/no-inline-comments.md) (already in lint-trap)
- [`no-mixed-requires`](https://github.com/eslint/eslint/blob/master/docs/rules/no-mixed-requires.md) (probably a noop since we require one var per line)
- [`no-throw-literal`](https://github.com/eslint/eslint/blob/master/docs/rules/no-throw-literal.md)
- [`no-unexpected-multiline`](https://github.com/eslint/eslint/blob/master/docs/rules/no-unexpected-multiline.md)
- [`no-unneeded-ternary`](https://github.com/eslint/eslint/blob/master/docs/rules/no-unneeded-ternary.md)
- [`array-bracket-spacing`](https://github.com/eslint/eslint/blob/master/docs/rules/array-bracket-spacing.md) (was already in lint-trap)
- [`block-scoped-var`](https://github.com/eslint/eslint/blob/master/docs/rules/block-scoped-var.md)
- [`computed-property-spacing`](https://github.com/eslint/eslint/blob/master/docs/rules/computed-property-spacing.md)
- [`dot-location`](https://github.com/eslint/eslint/blob/master/docs/rules/dot-location.md) (dot on property which is the most common style I've seen)
- [`newline-after-var`](https://github.com/eslint/eslint/blob/master/docs/rules/newline-after-var.md)
- [`object-curly-spacing`](https://github.com/eslint/eslint/blob/master/docs/rules/object-curly-spacing.md) (was already in lint-trap)
- [`operator-linebreak`](https://github.com/eslint/eslint/blob/master/docs/rules/operator-linebreak.md) (linebreaks go after infix operators)
- [`space-before-function-paren`](https://github.com/eslint/eslint/blob/master/docs/rules/space-before-function-paren.md)

@Raynos @lxe @mlmorg @johnmegahan @jcorbin @kriskowal

andrewdeandrade committed:
Non-controversial new eslint 0.24.0 rules and cleanup up formatting

andrewdeandrade committed:
Make operator-linebreak options explicit

johnmegahan commented:
:+1: lgtm

Raynos commented on rc/.eslintrc.json:
```diff
@@ -20,38 +20,38 @@
         "no-console": 2,
         "no-constant-condition": 2,
-        "no-continue": 0,
+        "no-continue": 2,
```
This is not the place to have opinions on how to write loops.

We don't have a `no-while` or `no-do-loop` statement either.

andrewdeandrade replied:
Cool. We'll discuss this one in a different PR.

Raynos commented on rc/.eslintrc.json:
```diff
@@ -20,38 +20,38 @@
         "no-extra-bind": 2,
         "no-extra-boolean-cast": 2,
-        "no-extra-parens": 0,
+        "no-extra-parens": 2,
```
This may hurt readability but let's give it a try.

andrewdeandrade replied:
OK

Raynos commented on rc/.eslintrc.json:
```diff
@@ -114,15 +117,15 @@
         "no-trailing-spaces": 2,
-        "no-this-before-super": 0,
-        "no-throw-literal": 0,
+        "no-this-before-super": 2,
```
this does not belong here; belongs in ES6 file.

andrewdeandrade replied:
I will remove.

Raynos commented on rc/.eslintrc.json:
```diff
@@ -114,15 +117,15 @@
         "no-undef-init": 2,
         "no-undefined": 0,
-        "no-unexpected-multiline": 0,
+        "no-unexpected-multiline": 2,
```
Nice :)

Raynos commented on rc/.eslintrc.json:
```diff
@@ -153,15 +156,21 @@
+        ],
         "accessor-pairs": 0,
-        "block-scoped-var": 0,
+        "block-scoped-var": 2,
```
opinionated; but a good thing.

andrewdeandrade replied:
We'll discuss this in a separate PR.

Raynos commented on rc/.eslintrc.json:
```diff
@@ -177,7 +186,10 @@
             11
         ],
-        "computed-property-spacing": [0, "never"],
+        "computed-property-spacing": [
```
Does not belong in this file; this is an ES6 thing.

Raynos replied:
Never mind; this applies to ES5 too; +1

andrewdeandrade replied:
OK

Raynos commented on rc/.eslintrc.json:
```diff
@@ -247,14 +267,23 @@
-        "object-curly-spacing": [0, "never"],
+        "object-curly-spacing": [
+            2,
+            "never"
```
i would argue this should be set to "always".

andrewdeandrade replied:
This was the default in lint-trap. I too prefer "always" here. I was going with the previous standard, but I'm fine to change this since I think it makes things more readable.

What about the array-spacing rule?

Raynos replied:
As for array-spacing. I would say no; arrays and functions are the same, they are N things.

Where as objects are key values seperated by new lines. When you put them on one line you seperate them by spaces instead.

mlmorg replied:
This is different than what was in lint-trap and now break linting in applications which all do `{foo: 'bar'}`. Can we revert this?

Raynos commented on rc/.eslintrc.json:
```diff
@@ -277,7 +312,10 @@
-        "space-before-function-paren": [0, "always"],
+        "space-before-function-paren": [
+            2,
+            "always"
```
This should definitely be "never".

andrewdeandrade replied:
I think this is a no-op anyways since we require functions to be named, but yes, you are correct, this should be "never"

Raynos commented:
Finished review.

andrewdeandrade committed:
Changes incorporating jakev's feedback

andrewdeandrade committed:
Dedupe es6 rules that are in es5 mode

andrewdeandrade commented:
- rebased
- disabled `no-continue`
- changed `space-before-function-paren` to "never"
- changed `object-curly-spacing` to "always"
- deduped es6 mode rules

Raynos commented:
lgtm.
